File: VB/MvpvmNavigation/Services/EditFormService.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports DevExpress.Mvvm
Imports MvpvmNavigation.Model

Namespace MvpvmNavigation.Services
    Friend Class EditFormService
        Inherits DialogDocumentManagerService

        Private ReadOnly viewModuleType As ModuleType
        Public Sub New(ByVal viewModuleType As ModuleType)
            Me.viewModuleType = viewModuleType
        End Sub

        Protected Overrides Function CreateDocumentCore(ByVal documentType As String, ByVal viewModel As Object, ByVal parentViewModel As Object, ByVal parameter As Object) As IDocument
            Dim moduleLocator = GetService(Of Services.IModuleLocator)(parentViewModel)
            Dim view As Object = moduleLocator.GetModule(viewModuleType)
            viewModel = EnsureViewModel(viewModel, parameter, parentViewModel, view)
            Return RegisterDocument(view, Function(form) New DialogDocument(Me, form, viewModel), Function() New EditForm(documentType))
        End Function

        Private Class EditForm
            Inherits Form

            Public Sub New(ByVal text As String)
                Me.Text = text
                StartPosition = FormStartPosition.CenterParent
            End Sub
        End Class
    End Class
End Namespace

File: CS/MvpvmNavigation/Services/EditFormService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using DevExpress.Mvvm;
using MvpvmNavigation.Model;

namespace MvpvmNavigation.Services {
    class EditFormService : DialogDocumentManagerService {
        readonly ModuleType viewModuleType;
        public EditFormService(ModuleType viewModuleType) {
            this.viewModuleType = viewModuleType;
        }

        protected override IDocument CreateDocumentCore(string documentType, object viewModel, object parentViewModel, object parameter) {
            var moduleLocator = GetService<Services.IModuleLocator>(parentViewModel);
            object view = moduleLocator.GetModule(viewModuleType);
            viewModel = EnsureViewModel(viewModel, parameter, parentViewModel, view);
            return RegisterDocument(view, (form) => new DialogDocument(this, form, viewModel), () => new EditForm(documentType));
        }

        class EditForm : Form {
            public EditForm(string text) {
                Text = text;
                StartPosition = FormStartPosition.CenterParent;
            }
        }
    }
}
